<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: --.--.2026

----



##**Бизнес-контекст**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Этап 1 Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).



## **Шаг 1. Загрузка данных**

- Импорт библиотек и загрузка датасета из внешнего источника.

---
### **Библиотеки**




In [1]:
#загруза библиотек

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn   kagglehub -q
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


### **Загрузка датасета**

In [2]:
#Загрузка датасета
path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
file_path = os.path.join(path, [x for x in os.listdir(path) if x.endswith('.csv')][0])

#Сохранение
df = pd.read_csv(file_path)



100%|██████████| 6.66M/6.66M [00:00<00:00, 123MB/s]

Extracting files...


**Итого:**

 - Установлены необходимые библиотеки и датасет сохранён в `df`

-----
----
----

## **Шаг 2. Общий осмотр данных**

Первичный анализ структуры данных включает оценку следующих параметров, таких как:

- размерность датасета
- типы признаков и их соответствие содержанию
- наличие пропусков и их доля в каждом поле
- основные статистические характеристики числовых переменных

In [3]:
# Основная информация
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [4]:
#Вывод первых 10 строк
df.head(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom
5,581475,12/9/2019,21705,Bag 500g Swirly Marbles,10.65,24,13069.0,United Kingdom
6,581475,12/9/2019,22118,Joy Wooden Block Letters,11.53,18,13069.0,United Kingdom
7,581475,12/9/2019,22119,Peace Wooden Block Letters,12.25,12,13069.0,United Kingdom
8,581475,12/9/2019,22217,T-Light Holder Hanging Lace,10.65,12,13069.0,United Kingdom
9,581475,12/9/2019,22216,T-Light Holder White Lace,10.55,24,13069.0,United Kingdom


In [5]:
#Осмотр случайных строк
df.sample(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
351667,553015,5/12/2019,21174,Pottering In The Shed Metal Sign,12.38,1,12748.0,United Kingdom
519340,537823,12/8/2018,22781,Gumball Magazine Rack,25.91,1,15823.0,United Kingdom
12158,580999,12/6/2019,21790,Vintage Snap Cards,6.19,1,17883.0,United Kingdom
35925,579196,11/28/2019,22563,Happy Stencil Craft,7.24,1,14096.0,United Kingdom
528100,537197,12/5/2018,84997D,Childrens Cutlery Polkadot Pink,14.09,6,12647.0,Germany
403327,C547836,3/27/2019,22423,Regency Cakestand 3 Tier,21.47,-9,15465.0,United Kingdom
468799,541845,1/23/2019,22457,Natural Slate Heart Chalkboard,13.27,1,15167.0,United Kingdom
29961,579406,11/29/2019,22667,Recipe Box Retrospot,6.19,12,13014.0,United Kingdom
419214,546420,3/13/2019,22804,Pink Hanging Heart T-Light Holder,13.27,2,14810.0,United Kingdom
131278,572058,10/20/2019,22804,Pink Hanging Heart T-Light Holder,13.27,3,18252.0,United Kingdom


In [6]:
#Осмотр по различным показателям
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


### **Результаты первичного осмотра**

**Общая информация:**
- Датасет содержит 536 350 записей, 8 столбцов.
- Пропуски: в поле `CustomerNo`.
- Типы данных: `Date` и `CustomerNo` требуют преобразования для дальнейшего анализа.
- `Price`: среднее — 12.66 руб., медиана — 11.94 руб. Диапазон: от 5.13 до 660.62 руб.
- `Quantity`: среднее — 9.92 шт., медиана — 3 шт. Диапазон: от -80 995 до 80 995 шт., что указывает на наличие экстремальных значений.
  
**Выводы:**
  
  -  Данные содержат аномалии в `Quantity` (отрицательные и экстремально высокие значения) и выбросы в `Price`.
  - Требуется преобразование типов и принятие решения по обработке аномалий. В остальном данные пригодны для анализа.
  - Требуется осомотр пропусков в поле `CustomerNo`.


## **Шаг 3. Обработка типов данных**

- проверка и приведение типов данных к форматам, соответствующим их содержанию
- обеспечение согласованности типов для последующих операций
- оптимизация структуры данных для дальнейшей работы
- `Date` преобразовать в `datetime` для временной аналитики
- `CustomerNo` привести  к `Int64` для корректной группировки клиентов

In [7]:
#Осмотр типов данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [8]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования**

Типы данных приведены к форматам, необходимым для дальнейшего анализа

- `Date` преобразована в `datetime`, что открывает доступ к исследованию временных рядов и сезонных срезов;
- `CustomerNo` переведён в `Int64` с поддержкой пропусков, что обеспечивает корректную идентификацию клиентов при группировках.

Остальные типы данных оставлены без изменений.

---
----

## **Шаг 4 Поиск и обработка пропусков**

- Проверка данных на наличие пропущенных значений
- оценка их объема
- анализ возможного влияния на результаты исследования и принятие
- решения по их обработке

In [9]:
# колличесвто и доля пропусков

df.isnull().sum(), df.isnull().mean()

(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [10]:
# отображение данных

df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


####**Описание и дальнейшие действия:**

* Пропуски обнаружены только в поле `CustomerNo` и составляют **55** записей (0.01%).
* Для сохранения всех наблюдений пропуски будут заменены значением `-1`, которое в дальнейшем будет обозначать неизвестного клиента

----

In [30]:
# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

# проверка после заполнения пропусков
display(df[df['CustomerNo'] == -1].head(10))

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity,Year,Month,DayOfWeek
231302,558245,2019-06-27,22734,Set Of 6 Ribbons Vintage Christmas,10.25,-1,United Kingdom,8,2019,6,3
518274,C538686,2018-12-14,22467,Gumball Coat Rack,12.86,-1,United Kingdom,-1,2018,12,4
518531,C539756,2018-12-21,22720,Set Of 3 Cake Tins Pantry Design,15.32,-1,United Kingdom,-1,2018,12,4
518548,C539948,2018-12-23,21888,Bingo Set,14.09,-1,United Kingdom,-4,2018,12,6
518575,C540155,2019-01-05,72802B,Ocean Scent Candle In Jewelled Box,14.16,-1,United Kingdom,-54,2019,1,5
518740,C540559,2019-01-10,21888,Bingo Set,14.09,-1,United Kingdom,-4,2019,1,3
518795,C540854,2019-01-12,22461,Savoy Art Deco Clock,23.32,-1,United Kingdom,-1,2019,1,5
519171,C542540,2019-01-28,21658,Glass Beurre Dish,18.75,-1,United Kingdom,-1,2019,1,0
519240,C542860,2019-02-01,22580,Advent Calendar Gingham Sack,16.35,-1,United Kingdom,-1,2019,2,4
519279,C543185,2019-02-04,22332,Skulls Party Bag + Sticker Set,11.94,-1,United Kingdom,-14,2019,2,0


### **Итог работы c пропусками:**


Пропуски в `CustomerNo` обработаны заменой на значение `-1`.


Пропущенных значений в датасете не осталось.



---
---

## **Шаг 5. Поиск и анализ  явных/неявных дубликатов**

- Выявление избыточных записей (полностью совпадающих строк)

- оценка их объёма и доли в общем массиве данных

- анализ возможного влияния дубликатов на агрегированные метрики

- принятие решения по обработке: удаление или агрегация

---

###**Анализ явных дубликатов:**

- поиск полностью совпадающих строк по всем колонкам
- оценка количества и доли дубликатов в данных
- анализ влияния на метрики (выручка, количество заказов, товаров)


In [12]:
# Количество и доля явных дубликатов

duplicates_count = df.duplicated().sum()
duplicates_percent = (duplicates_count / len(df)) * 100

print(f"Количество дубликатов: {duplicates_count}")
print(f"Процент дубликатов: {duplicates_percent:.2f}%")

Количество дубликатов: 5200
Процент дубликатов: 0.97%


In [13]:
#Вывод записей дубликатов

df[df.duplicated()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


####**Дальнейшие действия с явными дубликатами:**


Обнаружено **5200** дубликатов (0.97% от всех записей).  

Удаление не повлияет на общую структуру данных и исключит искажение метрик.

In [14]:
#Удаление дубликатов и проверка

df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

####**Итог обработки:**

Явные дубликаты полностью удалены.

 В датасете не осталось полностью совпадающих строк.


----
-----

### **Анализ неявных дубликатов:**

- Проверка наличия повторяющихся товаров в рамках одной транзакции.
- Оценка доли неявных дубликатов.
- Анализ структуры повторяющихся записей.
- Проверка возможности их объединения без потери информации.
- Выбор и обоснование обработки.

In [15]:
# Поиск неявных дубликатов

implicit_duplicates = (
    df.groupby(['TransactionNo', 'ProductNo'])
      .size()
      .reset_index(name='count')
      .query('count > 1')
)

display(implicit_duplicates.head(10))

print(f'Найдено комбинаций с повторяющимися товарами: {len(implicit_duplicates)}')



,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2


Найдено комбинаций с повторяющимися товарами: 5553


In [16]:
# Строки, относящиеся к найденным комбинациям

duplicate_rows = df.merge(
    implicit_duplicates[['TransactionNo', 'ProductNo']],
    on=['TransactionNo', 'ProductNo']
)

print(f'Количество строк с повторами: {len(duplicate_rows)}')
print(f'Доля строк от общего объема данных: {len(duplicate_rows) / len(df) * 100:.2f}%')

Количество строк с повторами: 11336
Доля строк от общего объема данных: 2.13%


In [17]:
# Пример одной транзакции с повторяющимися товарами

sample_transaction = implicit_duplicates['TransactionNo'].iat[0]

display(
    df.loc[
        df['TransactionNo'] == sample_transaction,
        ['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Price', 'Quantity']
    ].sort_values('ProductNo')
)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity
533393,536381,2018-12-01,15056N,Edwardian Parasol Natural,16.35,2
533379,536381,2018-12-01,21166,Cook With Wine Metal Sign,12.25,1
533378,536381,2018-12-01,21169,You're Confusing Me Metal Sign,11.98,3
533380,536381,2018-12-01,21175,Gin And Tonic Diet Metal Sign,12.40,2
533400,536381,2018-12-01,21523,Doormat Fancy Font Home Sweet Home,17.17,10
533391,536381,2018-12-01,21533,Retrospot Large Milk Jug,15.32,1
533392,536381,2018-12-01,21557,Set Of 6 Funky Beakers,13.27,2
533371,536381,2018-12-01,21672,White Spot Red Ceramic Drawer Knob,11.53,6
533397,536381,2018-12-01,21731,Red Toadstool Led Night Light,11.94,2
533377,536381,2018-12-01,21934,Skull Shoulder Bag,11.94,10


---
Также, для точности, необходимо будет:

- проверить корректность объединения данных

- убедиться, что повторяющиеся товары не имеют разницу по цене

In [18]:
price_check = (
    duplicate_rows
    .groupby(['TransactionNo', 'ProductNo'])
    .agg(unique_prices=('Price', 'nunique'))
    .query('unique_prices > 1')
)

print(f'Комбинаций с разными ценами: {len(price_check)}')

display(price_check.head())

Комбинаций с разными ценами: 1033


unique_prices
TransactionNo ProductNo               
536544        21218                  2
              22844                  2
              22862                  2
              22866                  2
536592        22165                  2

####**Рузультаты проверки и дальнейшие действия**

 В ходе проверки обнаружены две комбинации `TransactionNo` и `ProductNo`, в которых один и тот же товар имеет разные значения `Price`.

Объединение  по `TransactionNo` и `ProductNo` может привести к потере информации, это стоит учитывать
и поле `Price` включается в группировку.


---

 Отрицательные значения `Quantity `на данный момент не изменяются.

 Отрицательные значения `Quantity`  участвуют в объединении наравне с остальными записями, поскольку отражают отдельный тип операций (например, возвраты) и будут исследоваться далее отдельно.


 ---

Для сохранения данных,  позиции будут объединены по  `Quantity.`

Группировка выполняется по:

- `TransactionNo`
- `Date`
- `ProductNo`
- `ProductName`
- `Price`
- `CustomerNo`
- `Country`

In [19]:
# Объединение неявных дубликатов

df = (
    df.groupby(
        [
            'TransactionNo',
            'Date',
            'ProductNo',
            'ProductName',
            'Price',
            'CustomerNo',
            'Country'
        ],
        as_index=False
    )['Quantity']
    .sum()
)

print(f'Размер датафрейма после объединения: {df.shape}')

display(df.head())

Размер датафрейма после объединения: (526417, 8)


,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6


In [20]:
remaining_duplicates = (
    df.groupby(['TransactionNo', 'ProductNo', 'Price'])
      .size()
      .gt(1)
      .sum()
)

print(f'Оставшихся неявных дубликатов: {remaining_duplicates}')

Оставшихся неявных дубликатов: 0


####**Итоги**

- Неявные дубликаты успешно объединены.

- Повторяющиеся позиции одного товара внутри одной транзакции сведены в одну запись путем суммирования `Quantity.`

- Использование `Price` в ключе группировки позволило избежать объединения записей с различной стоимостью товара.

- Проверка показала, что после обработки неявные дубликаты отсутствуют.

- Датафрейм готов к дальнейшему исследованию.

*Общий вывыод по шагу 5*

-----
----
---

##**Шаг 6. Анализ особенностей данных**

Выявление аномалий, особенностей распределений и скрытых закономерностей, которые могут повлиять на дальнейший анализ.

В рамках данного шага будут рассмотрены:
- распределение `Price`
-  распределение `Quantity` (включая отрицательные значения)
- исследование выбросов и длинных хвостов  
- анализ уникальных значений  
- поиск дополнительных закономерностей и аномалий



###**Анализ распределение цены**

In [21]:
df['Price'].quantile([
    0.25,
    0.50,
    0.75,
    0.90,
    0.95,
    0.99
])

,Price
0.25,11.02
0.50,11.94
0.75,14.09
0.90,17.90
0.95,19.61
0.99,27.30


In [22]:
q99 = df['Price'].quantile(0.99)

print(df[df['Price'] > q99].shape[0])

4469


In [23]:
print(f"Минимальная цена: {df['Price'].min()}")
print(f"Максимальная цена: {df['Price'].max()}")

Минимальная цена: 5.13
Максимальная цена: 660.62


In [24]:
top_10_expensive_products = (
    df[['ProductName', 'Price']]
    .drop_duplicates()
    .sort_values(by='Price', ascending=False)
    .head(15)
)

top_10_expensive_products

,ProductName,Price
27486,Vintage Blue Kitchen Cabinet,660.62
4744,Vintage Red Kitchen Cabinet,660.62
80360,Vintage Red Kitchen Cabinet,594.50
130068,Vintage Blue Kitchen Cabinet,594.50
105952,Love Seat Antique White Metal,210.12
23930,Love Seat Antique White Metal,189.62
238,Rustic Seventeen Drawer Sideboard,179.37
18504,Regency Mirror With Shutters,179.37
18503,Rustic Seventeen Drawer Sideboard,158.88
83700,Regency Mirror With Shutters,158.88


----

**Проверить почему цена на один и тот же товор разная и узнать факторы которые на это повлияли но это потом сейчас фиксируем факт**

In [25]:
df[df['ProductName'] == 'Vintage Blue Kitchen Cabinet']

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
27486,538662,2018-12-13,22656,Vintage Blue Kitchen Cabinet,660.62,15159,United Kingdom,2
30086,538999,2018-12-15,22656,Vintage Blue Kitchen Cabinet,660.62,16003,United Kingdom,2
30614,539044,2018-12-15,22656,Vintage Blue Kitchen Cabinet,660.62,15727,United Kingdom,2
36754,539581,2018-12-20,22656,Vintage Blue Kitchen Cabinet,660.62,16143,United Kingdom,2
51949,540946,2019-01-12,22656,Vintage Blue Kitchen Cabinet,660.62,12359,Cyprus,1
80361,543370,2019-02-07,22656,Vintage Blue Kitchen Cabinet,660.62,12359,Cyprus,1
99647,545158,2019-02-28,22656,Vintage Blue Kitchen Cabinet,660.62,15874,United Kingdom,1
103707,545543,2019-03-03,22656,Vintage Blue Kitchen Cabinet,660.62,13476,United Kingdom,1
106631,545759,2019-03-07,22656,Vintage Blue Kitchen Cabinet,660.62,16316,United Kingdom,2
107941,545902,2019-03-07,22656,Vintage Blue Kitchen Cabinet,660.62,15834,United Kingdom,1


#### **Краткий вывод что я нашел и что дальше с этим делать**

### **Анализ высоких значений Price**

Ранее мы заметили аномально высокие значения в столбце `Price` (до `660.62` руб.), что может влиять на средние показатели и искажать анализ прибыльности. Мы уже проанализировали 99-й перцентиль (`27.30`), и обнаружили, что есть значения гораздо выше этого. Эти высокие цены могут принадлежать к товарам класса люкс, редким предметам или быть ошибками ввода данных.

**Предлагаемые действия:**

1.  **Исследовать товары с аномально высокими ценами**: Проверить, что это за товары и соответствуют ли их названия таким высоким ценам. Возможно, это ошибки в данных или особые категории товаров.
2.  **Принять решение о сохранении/удалении**: Если это валидные дорогие товары, их можно оставить. Если это ошибки, их следует скорректировать или удалить. Для большинства анализов, такие выбросы могут быть проблемными.

На данный момент, мы не будем удалять эти строки, так как они могут представлять ценные данные о дорогих товарах. Вместо этого, мы признаем их существование и будем учитывать при дальнейшем анализе. Если потребуется более глубокий анализ ценовых сегментов, можно будет вернуться к их обработке.

----
---
---

#!// **Анализ отриц значений**






In [26]:
display(df[df['Quantity'] < 0].head(15).sort_values(by='Quantity'))

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
517901,C536391,2018-12-01,21983,Pack Of 12 Blue Paisley Tissues,10.55,17548,United Kingdom,-24
517900,C536391,2018-12-01,21980,Pack Of 12 Red Retrospot Tissues,10.55,17548,United Kingdom,-24
517903,C536391,2018-12-01,22553,Plasters In Tin Skulls,11.94,17548,United Kingdom,-24
517902,C536391,2018-12-01,21984,Pack Of 12 Pink Paisley Tissues,10.55,17548,United Kingdom,-24
517904,C536391,2018-12-01,22556,Plasters In Tin Circus Parade,11.94,17548,United Kingdom,-12
517899,C536391,2018-12-01,21484,Chick Grey Hot Water Bottle,13.79,17548,United Kingdom,-12
517905,C536391,2018-12-01,22557,Plasters In Tin Vintage Paisley,11.94,17548,United Kingdom,-12
517906,C536506,2018-12-01,22960,Jam Making Set With Jars,14.61,17897,United Kingdom,-6
517912,C536548,2018-12-01,22077,6 Ribbons Rustic Charm,11.94,12472,Germany,-6
517911,C536548,2018-12-01,21218,Red Spotty Biscuit Tin,14.09,12472,Germany,-3


### **Анализ отрицательных значений Quantity**

Отрицательные значения в столбце `Quantity` (обнаружено `8548` записей, составляющих `1.61%` от общего числа) обычно указывают на возвраты товаров или отмененные заказы. Для целей анализа продаж и выручки, эти записи могут искажать общую картину, так как они представляют собой не продажи, а их отмену.

**Предлагаемые действия:**

1.  **Выделить возвраты**: Если `TransactionNo` начинается с 'C', это может прямо указывать на отмену (cancellation). Необходимо проверить это предположение.
2.  **Удалить отрицательные значения**: Для анализа чистых продаж, возвраты часто исключают из основного датафрейма. Их можно анализировать отдельно, если требуется изучить динамику возвратов.
3.  **Преобразовать в абсолютные значения**: Если цель — понять общий объем движения товаров, можно использовать абсолютные значения, но это не рекомендуется для анализа выручки.

В рамках данного EDA, мы будем считать, что отрицательные `Quantity` представляют собой возвраты/отмены и **удалим их из основного датафрейма**, чтобы сфокусироваться на чистых продажах.

### **Причина и обработка отрицательных значений Quantity**

Отрицательные значения в столбце `Quantity` обычно указывают на **возвраты товаров или отмененные заказы**. Это подтверждается тем, что все такие записи имеют `TransactionNo`, начинающийся с буквы 'C', что является стандартным обозначением для транзакций отмены или возврата.

**Дальнейшие действия:**

Для целей анализа продаж и выручки, эти записи были **удалены** из основного датафрейма (`df`). Это позволяет сфокусироваться исключительно на данных о состоявшихся продажах, избегая искажения метрик. Если потребуется анализ динамики возвратов, эти данные могут быть выделены и проанализированы отдельно.

----
----
---

### **Осмотр уникальных позиций**

In [27]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,23204
CustomerNo,4739
ProductName,3768
Country,38


### **Анализ хвостов и подсчет хваостов**

- нужно подсчитать что выше 99 квартиля и какая сумма формируется от общего число и понять выброс это или нет

In [28]:
df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,3.0
0.75,11.0
0.90,24.0
0.95,30.0
0.99,100.0


---
---
---

## **Шаг 7 Решение аномалий:**

#**//Общий вывод по первому этапу**

----
---
---

# **Этап 1.5 Решение данных которые вызывают сомнения**

----
----
----

#**Этап 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

### **Создание новых признаков из даты**

Для дальнейшего анализа сезонности и временных трендов, из колонки `Date` будут извлечены год, месяц и день недели.

In [29]:
# Извлечение года, месяца и дня недели
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['DayOfWeek'] = df['Date'].dt.dayofweek # Понедельник=0, Воскресенье=6

display(df.head())

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity,Year,Month,DayOfWeek
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6,2018,12,5
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2,2018,12,5
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6,2018,12,5
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6,2018,12,5
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6,2018,12,5


### **Агрегация данных на уровне заказа для анализа корзины**

Для анализа содержимого корзины и среднего чека, агрегируем данные по `TransactionNo`, чтобы получить общую сумму и количество уникальных товаров в каждом заказе.

### **Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен**

Давайте исследуем, какие товары являются наиболее прибыльными и популярными, а также как распределены цены на товары.

### **Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме**

Проанализируем структуру заказов, чтобы понять типичный размер корзины, средний чек и как распределяются заказы по их общей стоимости.

# **Этап 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

# **Этап 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.

----
----
----